Carthage - substrate v3

mnaamani opened the pull request:
Merges work from https://github.com/Joystream/joystream/pull/3875 with latest audit3 into carthage, with help of @ignazio-bovo and @Lezek123 resolved all conflicts.

Additionally migrated the project-token pallet https://github.com/Joystream/joystream/pull/3928/commits/8ac8fa6b1483dd8b295d6a016b0409bdd39b423e to substrate v3.

Build.
`yarn cargo-build`

Run a test node:
`./target/release/joystream-node --dev`

Although I've made sure to apply cargo-fmt, many clippy warnings need to be dealt with still. (so `yarn cargo-checks` will fail as well as docker build of joystream-node)

The bounty module was not migrated and is removed from the runtime. (Awaiting merge of bounties v2 module)

Several modules (content, proposals, storage, working-group) tests and benchmarks still need to be migrated, as well as the tests in the runtime crate.

One thing to note, for the staking pallet election configuration I went with the same configuration as the substrate node. Perhaps there is a simpler configuration using the phragmen pallet instead. Maybe that can be discussed.

iorveth committed:
Runtime: upgrade to substrate v3

iorveth committed:
Cargo fmt

iorveth committed:
Upgrade node & chainspec builder to the third version of substrate

iorveth committed:
Runtime: migrate benchmarking & unit tests to the substrate v3

iorveth committed:
Add Cargo.lock

iorveth committed:
Runtime: Upgrade weights

iorveth committed:
Runtime: convenient types for constant types

iorveth committed:
RocksDbWeight -> DbWeight

iorveth committed:
Undo *.ts files

iorveth committed:
Runtime: Restore original cli properties

iorveth committed:
Runtime: restore constant

iorveth committed:
Runtime: provide OffencesWeightSoftLimit parameter

iorveth committed:
Runtime: fix node tests

iorveth committed:
Runtime: remove redundant dependency

iorveth committed:
Substrate v3: restore .ts files

iorveth committed:
node_runtime.compact.wasm. ->  joystream_node_runtime.compact.wasm.

iorveth committed:
fix networks-tests .env

iorveth committed:
fix integration-tests .env

iorveth committed:
Blog: remove serde dependency

iorveth committed:
Bump crate versions

iorveth committed:
node/service: restore tests

iorveth committed:
Fix balances chainspec

iorveth committed:
Runtime: Move serde to dev dependencies

iorveth committed:
Runtime: remove redunant clone

iorveth committed:
Runtime: start node tests reworking

iorveth committed:
proposals_codex: remove redundant iheritance

iorveth committed:
/node: restore command defaults

iorveth committed:
Restore analyses/bench

iorveth committed:
Fix multile clippy errors

iorveth committed:
Runtime: resolve the rest of clippy issues

iorveth committed:
Runtime: fix /node tests & upgrade implementation

iorveth committed:
Node: upgrade cli & command implementation

iorveth committed:
Runtime: fix node tests issues, clippy

iorveth committed:
Runtime: fix compilation issue

iorveth committed:
Working-group: fix failing benchmark test

shamil-gadelshin committed:
Merge pull request #2558 from iorveth/olympia_audit

Substrate v3 upgrade

singulart committed:
https://github.com/Joystream/joystream/issues/3530 added test that illustrated my concerns

mnaamani committed:
fix initial balances and initial members args

mnaamani committed:
Merge pull request #3709 from mnaamani/master-devops-fix-create-chain-spec

fix initial balances and initial members args

Commit:
Distributor node: Remove dependency on readChunk

Commit:
Distributor node: Fix formatting

mnaamani committed:
update readme and sample config for aws script deployments

zeeshanakram3 committed:
Initial QN metaprotocol for comments and reactions (#3726)

zeeshanakram3 committed:
Comments and reactions flatten video property (#3710)

mnaamani committed:
Merge pull request #3725 from mnaamani/aws-cf-better-sample-cfg-and-readme

update readme and sample config for aws script deployments

mnaamani committed:
update to hydra v3.1.0-alpha.26

mnaamani committed:
docker compose config for rhodes faucet

Commit:
Argus: Factor out FSPromise class, bump&expose node version

mnaamani committed:
Merge pull request #3728 from mnaamani/rhodes-update-hydra

update to hydra v3.1.0-alpha.26

mnaamani committed:
Merge pull request #3723 from Lezek123/distributor-node-fix-mime-type

Distributor node: Fix mime-type detection

mnaamani committed:
Merge pull request #3730 from mnaamani/rhodes-faucet-config

docker compose config for rhodes faucet

mnaamani committed:
Merge pull request #3718 from Joystream/rhodes

Rhodes Release

zeeshanakram3 committed:
Comments-Reactions: add parentCommentAuthor field (#3740)

zeeshanakram3 committed:
Comments-Reactions: review followups (#3340)

kdembler committed:
Merge pull request #3751 from Joystream/master

dobertRowneySr committed:
add failing test on auction with member owned channel with no reward

dobertRowneySr committed:
all tests passing

dobertRowneySr committed:
add test for buy_now and accept_incoming offer
    case : NFT is member-channel owned and reward account is not set

dobertRowneySr committed:
fix doc line

dobertRowneySr committed:
use ensure_reward_account instead

dobertRowneySr committed:
refactor reward payment methods & review tests

dobertRowneySr committed:
renaming

dobertRowneySr committed:
using unchecked channel_by_id

mnaamani committed:
bump spec runtime to 7

mnaamani committed:
remove setting NFT values in on_runtime_upgrade

mnaamani committed:
update runtime-update ci check workflow

mnaamani committed:
basic rhodes update-1 ci checks

mnaamani committed:
clippy fix

mnaamani committed:
appropriate name for runtime update job in ci check

dobertRowneySr committed:
cargo fmt

mnaamani committed:
Merge remote-tracking branch 'ignazio/master_fix_memberChannelNftProceedsEffectivelyCredited' into rhodes-update-1-ci-checks

Commit:
Argus: `cache-control: max-age` header fix

mnaamani committed:
Merge pull request #3767 from Joystream/rhodes-update-1-ci-checks

Rhodes update 1 CI Checks

mnaamani committed:
runtime CHANGELOG

mnaamani committed:
changelog type fix and additional details

bwhm committed:
Merge pull request #3774 from mnaamani/update-runtime-changelog

runtime CHANGELOG

mnaamani committed:
Merge pull request #3769 from Lezek123/argus-fix-cache-max-age

Argus: `cache-control: max-age` header fix

mnaamani committed:
fix vote stake release checks

zeeshanakram3 committed:
Qn comments section tweaks (#3781)

Commit:
Remove cyclic dependency between membership and working_group pallet

mnaamani committed:
fix test case, add test case for ordering

mnaamani committed:
use total vote power when inserting into winners list

mnaamani committed:
add check to test failure to multi-reveal

mnaamani committed:
check if vote is revealed in can_reveal_vote()

mnaamani committed:
Update runtime-modules/referendum/src/tests.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

mnaamani committed:
Merge pull request #3801 from mnaamani/referendum-single-reveal

Fix mutli reveal bug

mnaamani committed:
Merge pull request #3797 from mnaamani/referendum-winners-calculation-2

Fix referendum winners calculation

mnaamani committed:
Update runtime-modules/council/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

mnaamani committed:
Merge branch 'rhodes-update-2' into vote-stake-unlock-fix

mnaamani committed:
comprehensive test cases for release vote stake

mnaamani committed:
Merge pull request #3783 from mnaamani/vote-stake-unlock-fix

Fix Vote Stake Unlocking

mnaamani committed:
query-node council mappings update election round on success at end of election

zeeshanakram3 committed:
QN - Metaprotocol Status Event Tweak (#3793)

mnaamani committed:
runtime spec 8

mnaamani committed:
update runtime upgrade workflow

mnaamani committed:
use productio runtime in runtime upgrade test

mnaamani committed:
Merge pull request #3795 from Lezek123/fix-membership-wg-cyclic-dep

Remove cyclic dependency between `membership` and `working_group` pallet

zeeshanakram3 committed:
bump @joystream/metadata-protobuf version

mnaamani committed:
Merge pull request #3831 from zeeshanakram3/comments-reactions

bump @joystream/metadata-protobuf version

mnaamani committed:
Update query-node/mappings/src/council.ts

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

mnaamani committed:
drop redundant call to conclude election

mnaamani committed:
substrate v3 update

zeeshanakram3 committed:
add reactionsAndRepliesCount field to Comment entity (#3843)

zeeshanakram3 committed:
handle deletion of video having comments reactions (#3860)

mnaamani committed:
Merge branch 'master' into olympia_audit2_update

mnaamani committed:
a few pallet-membership fixes

kdembler committed:
Merge pull request #3818 from Joystream/comments-reactions

mnaamani committed:
Merge pull request #3815 from mnaamani/qn-conclude-election-when-council-elected

QN Update election round on successful council election

mnaamani committed:
Merge pull request #3826 from mnaamani/ci-check-upgrade-with-production-rutime

use production runtime in runtime upgrade test

mnaamani committed:
final fixes membership pallet

mnaamani committed:
Merge branch 'master' into rhodes-substrate-v3-update

mnaamani committed:
drop dependency on working group from membership module

mnaamani committed:
fix membership tests

mnaamani committed:
cargo-fmt

mnaamani committed:
fix membership pallet benchmarks

mnaamani committed:
fix pallet staking-handler tests

mnaamani committed:
Merge branch 'rhodes-update-2' into rhodes-substrate-v3-update

mnaamani committed:
fix referendum

mnaamani committed:
fix council pallet

mnaamani committed:
avoid "let mut" in membership pallet

mnaamani committed:
cargo-fmt

mnaamani committed:
fix working group

mnaamani committed:
cargo-fmt

mnaamani committed:
fix forum

mnaamani committed:
cargo-fmt

mnaamani committed:
clippy

mnaamani committed:
fix pallet-storage

mnaamani committed:
content WIP

mnaamani committed:
more content fixes

mnaamani committed:
cargo fmt

mnaamani committed:
proposals discussion

mnaamani committed:
cargo-fmt

mnaamani committed:
pallet-proposals-engine

mnaamani committed:
fix blog

mnaamani committed:
fix proposals codex

mnaamani committed:
disable bounty module

mnaamani committed:
fix runtime WIP

mnaamani committed:
fixes for runtime

mnaamani committed:
finished migrating runtime

mnaamani committed:
Merge pull request #3772 from Joystream/rhodes-update-1

Rhodes update 1

mnaamani committed:
completed node migration to substrate v3

mnaamani committed:
fix node tests

mnaamani committed:
add joystream pallets to dispatch_benchmark()

mnaamani committed:
updated chain-spec-builder

mnaamani committed:
Merge branch 'rhodes-substrate-upgrade' into rhodes-substrate-v3-update

mnaamani committed:
constitution: fix tests and benchmarks

mnaamani committed:
disable weight generation in ci, run benchmark tests only

mnaamani committed:
council: fix tests and benchmarks

mnaamani committed:
forum: fix tests

mnaamani committed:
forum: fix benchmarks

mnaamani committed:
Merge branch 'rhodes-substrate-v3-update' into carthage

mnaamani committed:
cargo-fmt and some missed merge conflict fixes

mnaamani committed:
content and proposals additional migration fixes

mnaamani committed:
migrate project-tokens pallet to substrate v3

mnaamani committed:
update node and runtime

mnaamani committed:
set final token decimal and ss58 prefix for joystream

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -39,12 +41,12 @@ pub use events::{Event, RawEvent};
-    type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+    type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
 
     // TODO: Add frame_support::pallet_prelude::TypeInfo trait
```
I think this comment can be removed now

Lezek123 commented on node/Cargo.toml:
```diff
@@ -3,7 +3,7 @@ authors = ['Joystream contributors']
 edition = '2018'
 name = 'joystream-node'
-version = '6.6.0'
+version = '7.1.0'
```
Shouldn't it be `7.0.0`?

mnaamani replied:
Yeah that doesn't look correct, although I normally try to match the minor number with the spec version so even that doesn't match.

Lezek123 commented on node/src/chain_spec/mod.rs:
```diff
@@ -21,46 +21,70 @@
 // Remove after the Antioch release.
```
Should be removed now?

mnaamani replied:
I'll have to run clippy and see if it can be removed.

Lezek123 commented on runtime-modules/blog/src/tests.rs:
```diff
@@ -1074,7 +1074,7 @@ fn reply_delete_success() {
         ));
 
-        assert_event_success(reply_created_event, number_of_events_before_call + 2)
+        assert_event_success(reply_created_event, number_of_events_before_call + 3)
```
Why was this value changed?

mnaamani replied:
Not sure, it was probably from merge, but since we are dropping blog pallet completely, I will not deal with it.

Lezek123 commented on runtime-modules/content/src/nft/mod.rs:
```diff
@@ -384,13 +384,24 @@ impl<T: Trait> Module<T> {
+    ///      - `Member` -> use member controller account
+    ///      - `CuratorGroup` -> Error
+    /// In order to statically guarantee that `video.in_channel` exists, by leveraging the
+    /// Runtime invariant: `video` exists => `video.in_channel` exists
```
The comment is now invalid, as there is no optional channel reward account.

mnaamani replied:
Will update comment to:
```
/// NFT owned by:
/// - Member: member controller account is used
/// - Channel: channel account
/// In order to statically guarantee that `video.in_channel` exists, by leveraging the
/// Runtime invariant: `video` exists => `video.in_channel` exists
```

Lezek123 commented on runtime-modules/content/src/permissions/curator_group.rs:
```diff
@@ -73,12 +73,14 @@ pub enum ContentModerationAction {
 pub type ModerationPermissionsByLevel<T> =
-    BTreeMap<<T as Trait>::ChannelPrivilegeLevel, BTreeSet<ContentModerationAction>>;
+    BTreeMap<<T as Config>::ChannelPrivilegeLevel, BTreeSet<ContentModerationAction>>;
+use scale_info::TypeInfo;
```
Shouldn't this `use` statement be at the beginning of the file?

mnaamani replied:
moved it to top, not sure why cargo-fmt didn't fix that

Lezek123 commented on runtime-modules/content/src/tests/nft/accept_incoming_offer.rs:
```diff
@@ -182,43 +183,49 @@ fn accept_incoming_offer_no_incoming_offers() {
             ContentActor::Member(DEFAULT_MEMBER_ID),
             video_id,
             NftIssuanceParameters::<Test>::default(),
         ));
```
Now I remembered why I refactored most of the setups to use `ContentTest`.
It was because storage buckets need to be initialized before channel is created + in general it's just a lot of boilerplate to create a channel, video and nft. I suggest using:
```suggestion
        ContentTest::with_member_channel().with_video_nft().setup();
```

Lezek123 commented on runtime-modules/content/src/tests/nft/buy_nft.rs:
```diff
@@ -426,16 +426,17 @@ fn buy_now_ok_with_nft_owner_curator_channel_correctly_credited() {
             video_id,
             SECOND_MEMBER_ID,
-            DEFAULT_NFT_PRICE,
+            100u64,
```
Probably `DEFAULT_NFT_PRICE` should be used instead of `100u64` (see the assertion below)

mnaamani replied:
will do

Lezek123 commented on runtime-modules/content/src/tests/nft/pick_open_auction_winner.rs:
```diff
@@ -661,3 +661,71 @@ fn pick_open_auction_fails_during_channel_transfer() {
 }
+
+#[test]
+fn auction_proceeds_ok_in_case_of_member_owned_channel_with_no_reward_account_is_auctioneer() {
```
This test should be removed, since it's testing no-reward-account case (which is now impossible) and is bound to fail.

mnaamani replied:
This was removed by @ignazio-bovo in https://github.com/mnaamani/joystream/pull/12

Lezek123 commented on runtime-modules/content/src/types.rs:
```diff
@@ -96,11 +96,13 @@ impl<BlockNumber: BaseArithmetic + Copy> NftCounter<BlockNumber> {
     }
 }
+use frame_support::PalletId;
+use scale_info::TypeInfo;
```
Shouldn't the `use` statements be at the top of the file? (I think this may be the case in a few other places too, due to conflict resolutions)

mnaamani replied:
`cargo-fmt` re-formats these imports with alphabetic order, I think that is why it appears there?

mnaamani replied:
Weird, I moved it up

Lezek123 commented on runtime-modules/council/src/benchmarking.rs:
```diff
@@ -817,81 +816,81 @@ mod tests {
         let config = default_genesis_config();
         build_test_externalities(config).execute_with(|| {
-            assert_ok!(test_benchmark_announce_candidacy::<Runtime>());
+            assert_ok!(Council::<Runtime>::test_benchmark_announce_candidacy());
```
Should we replace those with https://crates.parity.io/pallet_staking/benchmarking/macro.impl_benchmark_test_suite.html?

mnaamani replied:
yes these fixes are on the the update-tests PR

mnaamani replied:
I've used that macro in the membership benchmarking. Its nice, but you can only use it with a single way of setting up externalities. While I was migrating other pallets I found some tests use different initial setups so just to save time I didn't bother using it everywhere, in case we later needed later to have different setups.

Lezek123 commented on runtime/Cargo.toml:
```diff
@@ -4,63 +4,72 @@ edition = '2018'
 lite-json = { version = '0.1.3', default-features = false}
-codec = { package = 'parity-scale-codec', version = '1.3.4', default-features = false, features = ['derive'] }
+codec = { package = 'parity-scale-codec', version = '3.1.2', default-features = false, features = ['derive'] }
+scale-info = { version = "2.1.1", default-features = false, features = ["derive"] }
```
Missing `scale-info/std` in `std` features

mnaamani replied:
added. surprised I didn't get any compiler errors though.

Lezek123 commented on runtime-modules/membership/src/lib.rs:
```diff
@@ -145,10 +148,9 @@ pub trait Trait:
-pub type Membership<T> = MembershipObject<<T as frame_system::Trait>::AccountId>;
-
-#[derive(Encode, PartialEq, Decode, Debug, Default)]
+#[derive(Encode, PartialEq, Decode, Debug, TypeInfo, Clone)]
```
This is the only place in the code where `TypeInfo` is not the last derived trait, which could be slightly confusing:
```suggestion
#[derive(Encode, Decode, PartialEq, Debug, Clone, TypeInfo)]
```

mnaamani replied:
done

Lezek123 commented on runtime-modules/membership/src/lib.rs:
```diff
@@ -511,17 +513,19 @@ decl_module! {
+                });
             }
 
+            // handle may not have been updated if not unique!
```
I don't think that's possible because of `transpose`?

dobertRowneySr replied:
Correct: `get_handle_hash` makes sure that if parameter `handle` is some then its uniqueness is checked.
if not `get_handle_hash` throws and `transpose?` will catch that.

mnaamani replied:
Thank you for clarifying, removing the comment.

mnaamani committed:
fix runtime tests

mnaamani commented:
I'm updating the tests in a separate PR https://github.com/mnaamani/joystream/pull/11

Lezek123 commented on runtime-modules/proposals/codex/src/tests/mock.rs:
```diff
@@ -3,38 +3,53 @@
+        ProposalsCodex: proposals_codex::{Module, Call, Storage, Event<T>},
+        ProposalsEngine: proposals_engine::{Module, Call, Storage, Event<T>},
+        Council: council::{Module, Call, Storage, Event<T>},
+        Referendum: referendum::<Instance0>::{Module, Call, Storage, Event<T>},
```
In `council/src/mock.rs` we have `referendum::<Instance1>`, shouldn't it be the same one here?

mnaamani replied:
Yep these fixes I'm working on in the update-tests PR

mnaamani replied:
Fixed as part of tests update https://github.com/mnaamani/joystream/pull/11

Lezek123 commented on runtime-modules/referendum/src/lib.rs:
```diff
@@ -29,6 +29,8 @@
 /////////////////// Configuration //////////////////////////////////////////////
 #![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::unused_unit)]
+#![allow(clippy::result_unit_err)]
```
It seems like there's only one occurrence of this warning, which looks like an easy fix, perhaps you could fix it?

mnaamani replied:
I'm planning another PR to fix all clippy errors/warnings https://github.com/Joystream/joystream/issues/3948

Lezek123 commented on runtime-modules/storage/Cargo.toml:
```diff
@@ -1,32 +1,34 @@
-randomness-collective-flip = { package = 'pallet-randomness-collective-flip', default-features = false, git = 'https://github.com/paritytech/substrate.git', rev = '2cd20966cc09b059817c3ebe12fc130cdd850d62'}
+sp-io = { package = 'sp-io', default-features = false, git = 'https://github.com/paritytech/substrate.git', rev = '6cbe1772bf258793fa9845daa8f43ea0cadee596'}
+sp-core = { package = 'sp-core', default-features = false, git = 'https://github.com/paritytech/substrate.git', rev = '6cbe1772bf258793fa9845daa8f43ea0cadee596'}
+balances = { package = 'pallet-balances', default-features = false, git = 'https://github.com/paritytech/substrate.git', rev = '6cbe1772bf258793fa9845daa8f43ea0cadee596'}
```
Looks like a duplicate, since it's already a main dependency

mnaamani replied:
I believe this was fixed in test update PR

Lezek123 replied:
It's still there

Lezek123 commented on runtime-modules/storage/Cargo.toml:
```diff
@@ -1,32 +1,34 @@
 [package]
 name = 'pallet-storage'
-version = '4.0.1'
+version = '5.0.0'
```
I'm not sure why some package versions were updated, while some others weren't (for example: `pallet-membership`)

mnaamani replied:
Not sure as well, they didn't come up in the merge conflict fixes, so I think there were inconsistent changes over the multiple branches we combined work from, so we should probably later just go through all crates and give them proper versions.

Lezek123 commented on runtime-modules/utility/src/lib.rs:
```diff
@@ -146,20 +150,21 @@ decl_module! {
         /// However, we treat this as a full block as `frame_system::Module::set_code` does
         /// # </weight>
-        #[weight = (T::MaximumBlockWeight::get(), DispatchClass::Operational)]
+        #[weight = (T::BlockWeights::get().get(DispatchClass::Operational).base_extrinsic, DispatchClass::Operational)]
```
`T::BlockWeights::get().get(DispatchClass::Operational).base_extrinsic` seems to be very different from `T::MaximumBlockWeight::get()`. Was it changed because we're now using `DispatchResultWithPostInfo`?

mnaamani replied:
Yes this does not look correct. As I remember from old substrate v2 that the weight for set_code() should be the maximum block weight. I'll check the substrate code base.

mnaamani replied:
I think we should use the same as this:
https://github.com/paritytech/substrate/blob/16bae92b3f257f9c15525ae3d0a23d036bd59715/frame/system/src/lib.rs#L410

mnaamani replied:
`#[weight = (T::BlockWeights::get().max_block, DispatchClass::Operational)]`

Lezek123 commented on runtime-modules/utility/src/tests/mocks.rs:
```diff
@@ -166,36 +160,56 @@ pub type OperationsWorkingGroupInstanceGamma = working_group::Instance8;
+        Timestamp: pallet_timestamp::{Module, Call, Storage, Inherent},
+        Utility: utility::{Module, Call, Event<T>},
+        Council: council::{Module, Call, Storage, Event<T>, Config<T>},
+        Referendum: referendum::<Instance0>::{Module, Call, Storage, Event<T>, Config<T>},
```
Same issue with `Instance0`/`Instance1` that I already pointed in one of the previous comments

mnaamani replied:
fixed in tests update PR

mnaamani committed:
council: fix tests

mnaamani committed:
referendum: fix tests

mnaamani committed:
working-group: fix tests

mnaamani committed:
constitution: fix benchmarks

mnaamani committed:
working-group: fix benchmarks

mnaamani committed:
membership: fix tests

Lezek123 commented on runtime/CHANGELOG.md:
```diff
@@ -1,3 +1,36 @@
+### Version 10.8.0 - upgrade
```
Missing entry for Substrate v3 upgrade

Lezek123 replied:
This is still missing, are you planning to add it in this PR?

Lezek123 commented on runtime/Cargo.toml:
```diff
@@ -4,63 +4,72 @@ edition = '2018'
 # Follow convention: https://github.com/Joystream/substrate-runtime-joystream/issues/1
 # {Authoring}.{Spec}.{Impl} of the RuntimeVersion
-version = '10.6.0'
+version = '11.20.0'
```
Why are we going from `10.6` to `11.20`? (instead of `11.0`?)

mnaamani replied:
I noticed recently that polkadot and substrate always increment the spec version of the runtime, never going back to 0 even when bumping the authoring version. I thought it might be best to follow the same practice. Although documentation clearly says that a validator will never attempt to produce blocks if its runtime doesn't match the authoring version, its doesn't definitely talk about regular nodes. But there is an issue perhaps with transaction replay across chains (signature checks cover only spec version not authoring version)

Lezek123 commented on runtime/src/lib.rs:
```diff
@@ -117,66 +140,100 @@ pub fn native_version() -> NativeVersion {
-    pub const MaximumBlockWeight: Weight = 2 * frame_support::weights::constants::WEIGHT_PER_SECOND;
-    pub const AvailableBlockRatio: Perbill = Perbill::from_percent(75);
-    pub const MaximumBlockLength: u32 = 5 * 1024 * 1024;
+    pub const BlockHashCount: BlockNumber = 2400;
```
What is the rationale behind choosing this number?

mnaamani replied:
I think I just took this from substrate node. It represents the number of blocks in the past an non-archive node will keep full storage state ?

mnaamani replied:
Although we have always had value of 250. Perhaps polkadot-js app needs a node to have more saved state to display some stats more accurately?

Lezek123 commented on runtime/src/lib.rs:
```diff
@@ -296,24 +363,32 @@ impl OnUnbalanced<NegativeImbalance> for DealWithFees {
 
 parameter_types! {
-    pub const UncleGenerations: BlockNumber = 0;
+    pub const UncleGenerations: BlockNumber = 5;
```
How was this number choosen?

mnaamani replied:
This came from substrate config. Will need to re-visit docs to see how this value affects block generation/forks etc.
We have always used 0 and so do polkadot and kusama.

Interestingly Polkadot uses 0 - https://github.com/paritytech/polkadot/blob/566a403be1277612f413136d9febe5efaabcf417/runtime/polkadot/src/lib.rs#L392

https://github.com/paritytech/polkadot/blob/566a403be1277612f413136d9febe5efaabcf417/runtime/kusama/src/lib.rs#L349

So I will revert to 0.

mnaamani committed:
content: update tests - some test failing

Lezek123 commented on runtime/src/runtime_api.rs:
```diff
@@ -258,29 +286,80 @@ impl_runtime_apis! {
+            use frame_benchmarking::{baseline, Benchmarking, BenchmarkList};
+            use frame_support::traits::StorageInfoTrait;
+
+            // Trying to add benchmarks directly to the Session Pallet caused cyclic dependency
+            // issues. To get around that, we separated the Session benchmarks into its own crate,
+            // which is why we need these two lines below.
+            use pallet_session_benchmarking::Pallet as SessionBench;
+            use pallet_offences_benchmarking::Pallet as OffencesBench;
+            use pallet_election_provider_support_benchmarking::Pallet as EPSBench;
```
I can't exactly parse what's the issue with cyclic dependencies here

mnaamani replied:
Neither could I, these comments came from substrate node code base.

Lezek123 replied:
Ok, initially I couldn't find the exact branch and repository where they originally appear, but it looks like they're here: https://github.com/paritytech/polkadot/blob/v0.9.23/runtime/polkadot/src/lib.rs#L1970 :+1:

Lezek123 commented on node/src/chain_spec/mod.rs:
```diff
@@ -208,57 +168,69 @@ pub fn testnet_genesis(
+            use rand::{seq::SliceRandom, Rng};
+            let limit = (MaxNominations::get() as usize).min(initial_authorities.len());
+            let count = rng.gen::<usize>() % limit;
+            let nominations = initial_authorities
```
This can generate empty set of nominations and cause the runtime to fail with:
```
Thread 'main' panicked at 'Expected Ok(_). Got Err(
    Module(
        ModuleError {
            index: 8,
            error: [
                4,
                0,
                0,
                0,
            ],
            message: Some(
                "EmptyTargets",
            ),
        },
    ),
)'
```

(tested with both 1 and 2 authorities)

Lezek123 reviewed (changes requested):
1. Most of the tests don't work (even the membership module ones), but I skipped it in this review, as this is beeing fixed in a separate PR. The only tests I have been able to run successfully were `project_token` tests and they seem to have actually now spotted a new issue (integer underflow)
2. I haven't been able to test any benchmarks yet either
3. The node builds successfully (although with a lot of clippy issues remaining)
4. I haven't been able to find any exposed type that's missing `TypeInfo`, metadata v14 also seems to be correctly generated
5. There are some remaining references to outdated rust toolchain:
```
5 results - 4 files

joystream-node-armv7.Dockerfile:
  2  RUN rustup self update
  3: RUN rustup install nightly-2021-02-20 --force
  4  RUN rustup default nightly-2021-02-20
  5  RUN rustup target add wasm32-unknown-unknown --toolchain nightly-2021-02-20
  6: RUN rustup component add --toolchain nightly-2021-02-20 clippy
  7  RUN apt-get update && \

node/README.md:
  59  ```bash
  60: cargo +nightly-2021-02-20 test --release --all
  61  ```

scripts/cargo-tests-with-networking.sh:
  11  echo 'running all cargo tests'
  12: cargo +nightly-2021-02-20 test --release --features "${FEATURES}" --all -- --ignored

scripts/raspberry-cross-build.sh:
  14      --volume ${HOME}/.cargo/registry:/home/cross/.cargo/registry \
  15:     joystream/rust-raspberry:nightly-2021-02-20 \
  16      build --release -p joystream-node
```
6. Trying to submit any extrinsic through https://polkadot.js.org/apps results in a following error:
```
signAndSend: error: l: -32602: Unknown SS58 address format `126`. ` `To support this address format, you need to call `set_default_ss58_version` at node start up.
```
7. The https://polkadot.js.org/apps also has trouble parsing our esistential deposit, which is below 1 `nanoJOY`. Since there is no unit < `nano`, it initially just displays an empty box:
  ![image](https://user-images.githubusercontent.com/19978260/175058536-b96910eb-2873-46d6-9828-6c144346255b.png)
  which then incorrectly changes to `nano` once I click on it:
  ![image](https://user-images.githubusercontent.com/19978260/175058675-9d1d4738-dc88-4cb5-9d11-921aa5c6fa01.png)

8. Some commands produce unexpected results when compared to substrate template node, for example:
  Template node:
  ```
  $ ./target/release/node-template benchmark machine
  2022-06-21 17:39:30 Running machine benchmarks... 
  ```
  Our node:
  ```
  $ ./target/release/joystream-node benchmark machine
  Error: Input("Please specify which chain you want to run, e.g. --dev or --chain=local")
  ```
9. I had a very hard time parsing through changes in `runtime/src` (mostly w.r.t. `pallet_election_provider_multi_phase`). I see that most of it is based on the substrate node configuration, but since I still don't understand this new election model fully, I can't really tell whether it makes sense.

mnaamani commented:
> I'm updating the tests in a separate PR [mnaamani#11](https://github.com/mnaamani/joystream/pull/11)

@ignazio-bovo I suspect there is something wrong with the NFT implementation after the merge.. or maybe I didn't update the tests correctly? see: https://github.com/mnaamani/joystream/pull/11#issuecomment-1163277162

mnaamani committed:
storage: partial fix tests

mnaamani committed:
storage: fix benchmarks

mnaamani committed:
storage: fix more tests

dobertRowneySr committed:
buyNow with nft member owner correctly credited fixed

dobertRowneySr committed:
fix buy now failing tests

dobertRowneySr committed:
accept incoming offer tests fixed

dobertRowneySr committed:
fix event emission assertions

dobertRowneySr committed:
add test for open auction payments

dobertRowneySr committed:
cargo fmt

dobertRowneySr commented on runtime-modules/working-group/src/lib.rs:
```diff
@@ -1553,19 +1558,24 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
                 };
 
-                Self::update_worker_missed_reward(worker_id, new_missed_reward);
+                Self::update_worker_missed_reward(worker_id, worker.clone(), new_missed_reward);
```
```suggestion
                Self::update_worker_missed_reward(worker_id, worker.to_owned(), new_missed_reward);
```

mnaamani replied:
will apply

dobertRowneySr commented on runtime-modules/working-group/src/lib.rs:
```diff
@@ -1580,7 +1590,7 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
         let new_missed_reward = missed_reward_so_far + reward;
 
-        Self::update_worker_missed_reward(worker_id, Some(new_missed_reward));
+        Self::update_worker_missed_reward(worker_id, worker.clone(), Some(new_missed_reward));
```
```suggestion
        Self::update_worker_missed_reward(worker_id, worker.to_owned(), Some(new_missed_reward));
```

mnaamani replied:
will apply

mnaamani committed:
impl IntegrityTest for working-group

mnaamani committed:
storage: fix all tests and benchmarks

mnaamani committed:
proposals-discussion: fix tests and benchmarks

mnaamani committed:
proposals-engine: fix test and benchmarks but some still failing

dobertRowneySr reviewed (changes requested):
- Logic in content pallet appears to be ok, since I have already fixed the tests and there were no problems wrt the implementation
- I have left some minor comments

mnaamani committed:
missing std feature on working group!

mnaamani committed:
council benchmarks: fix one name

mnaamani committed:
fix proposals-codex tests and benchmarks

mnaamani committed:
ignore some broken tests

mnaamani committed:
missed one fix in forum tests

mnaamani committed:
ignore all blog tests

mnaamani committed:
Merge remote-tracking branch 'ignazio/carthage_substrateV3__update_tests_ignazio' into carthage-substrate-v3-update-tests

mnaamani committed:
missed checking in changes to cargo toml file

mnaamani committed:
fix pallet-utility tests and benchmarks

mnaamani committed:
proposals engine fix remaining tests

mnaamani committed:
some clippy error fixes

mnaamani committed:
Merge pull request #11 from mnaamani/carthage-substrate-v3-update-tests

Update Tests

dobertRowneySr committed:
set initial next_sale_id value to 1

This is needed for the case Token is initialized in Idle state and
`update_upcoming_sale` is then called leading to underflow if the
initial value was 0

mnaamani committed:
update references to old toolchain

dobertRowneySr committed:
cleaner fix

mnaamani committed:
review request changes

mnaamani committed:
Merge remote-tracking branch 'ignazio/carthage__substrateV3_fix_token_sale_tests' into carthage-substrate-v3

mnaamani committed:
applied review feedback suggestions

mnaamani commented:
Pushed all the test fixes, including Ignazio's test fixes.

I've applied most of the requested changes in https://github.com/Joystream/joystream/pull/3928/commits/cf1f6191ef3d208c59727aa31b09968a38a3052e
and updated crate versions in https://github.com/Joystream/joystream/pull/3928/commits/22be7e2c3e816dbb274040eb70fcb6b1d7103468

Tx's work again in polkadot-js with the custom prefix:
<img width="683" alt="Screen Shot 2022-06-24 at 4 43 22 PM" src="https://user-images.githubusercontent.com/1621012/175542808-4c5d45dd-d09e-45b8-a924-b55944be3ceb.png">


Remaining Issues to address
- setting proper balances value that make sense, including the existential deposit.
- Look into chainspec construction https://github.com/Joystream/joystream/pull/3928#discussion_r903818603 it doesn't seem to be a problem when running dev chain for now.
- And of course other clippy warnings https://github.com/Joystream/joystream/issues/3948

Lezek123 commented on runtime-modules/storage/src/tests/mocks.rs:
```diff
@@ -2,62 +2,113 @@
+// The storage working group instance alias.
+pub type StorageWorkingGroupInstance = working_group::Instance2;
+// The distribution working group instance alias.
+pub type DistributionWorkingGroupInstance = working_group::Instance3;
```
In both `runtime/src/lib.rs` and `runtime-modules/utility/src/tests/mocks.ts` we have:
```suggestion
pub type DistributionWorkingGroupInstance = working_group::Instance9;
```

mnaamani replied:
They do not need to match as long as we are consistent in the specific constructed runtime.

mnaamani replied:
You might be right after all, as the benchmark command doesn't use the mock runtime, so it does make sense to use same instance number. Will update.

Lezek123 commented on runtime-modules/storage/src/tests/mocks.rs:
```diff
@@ -2,62 +2,113 @@
+        Storage: crate::{Pallet, Call, Storage, Config, Event<T>},
+        // Need to be added for benchmarks to work
+        Wg2: working_group::<Instance2>::{Pallet, Call, Storage, Event<T, I>},
+        Wg3: working_group::<Instance3>::{Pallet, Call, Storage, Event<T, I>},
```
```suggestion
        Wg9: working_group::<Instance9>::{Pallet, Call, Storage, Event<T, I>},
```

Lezek123 commented on runtime-modules/storage/src/benchmarking.rs:
```diff
@@ -15,38 +15,38 @@ use sp_std::iter::FromIterator;
 pub type StorageWorkingGroupInstance = working_group::Instance2;
 // The distribution working group instance alias.
-pub type DistributionWorkingGroupInstance = working_group::Instance9;
+pub type DistributionWorkingGroupInstance = working_group::Instance3;
```
```suggestion
pub type DistributionWorkingGroupInstance = working_group::Instance9;
```

mnaamani committed:
update joystream-node workflow with proper env variable set for profiles

Lezek123 commented on runtime-modules/content/src/tests/nft/accept_incoming_offer.rs:
```diff
@@ -181,73 +169,57 @@ fn accept_incoming_offer_no_incoming_offers() {
         ));
 
-        // Failure checked
+        // check creator owner balance increased by NFT PRICE + ROYALTY - ROYALTY - FEE
```
I don't understand why the comment states it's `NFT PRICE + ROYALTY - ROYALTY - FEE` instead of just `NFT PRICE - PLATFORM_FEE`

mnaamani replied:
Is this a case where implementation changed in one of our branches but comment came from another branch or just failed to update the comment when implementation changed?

mnaamani replied:
Fixed the comment.

dobertRowneySr replied:
Any seller receives: `nft_price - royalty - platform_fee`
Any creator receives after any nft exchange operation: `royalty`
if the seller is also the creator: `nft_price - royalty + royalty - platform_fee = nft_price - platform_fee`

Lezek123 commented on runtime-modules/content/src/tests/nft/pick_open_auction_winner.rs:
```diff
@@ -539,111 +526,73 @@ fn pick_open_auction_winner_fails_with_invalid_bid_commit() {
         ));
 
-        assert_eq!(channel_reward_account_balance(1u64), bid - platform_fee);
+        // net revenue for creator owner : NFT PRICE + ROYALTY - ROYALTY - FEE
```
Same comment as in [runtime-modules/content/src/tests/nft/accept_incoming_offer.rs](https://github.com/Joystream/joystream/pull/3928/files#diff-2df81e14936859a9f7980825a3a6a59d5633c12b1c564fc6c9db292fefb5ce04)

mnaamani replied:
Fixed comment

Lezek123 commented on runtime-modules/content/src/tests/fixtures.rs:
```diff
@@ -3009,6 +3011,16 @@ impl IssueNftFixture {
+            nft_metadata: self.params.nft_metadata.clone(),
+            non_channel_owner: Some(owner),
+            royalty: self.params.royalty.clone(),
+        };
```
```suggestion
        let new_params = NftIssuanceParameters::<Test> {
            non_channel_owner: Some(owner),
            ..self.params.clone()
        };
```

Lezek123 commented on runtime-modules/content/src/tests/fixtures.rs:
```diff
@@ -3017,12 +3029,34 @@ impl IssueNftFixture {
+            nft_metadata: self.params.nft_metadata.clone(),
+            non_channel_owner: self.params.non_channel_owner.clone(),
+            royalty: self.params.royalty.clone(),
+        };
```
```suggestion
        let new_params = NftIssuanceParameters::<Test> {
            init_transactional_status,
            ...self.params.clone()
        };

Lezek123 commented on runtime-modules/content/src/tests/fixtures.rs:
```diff
@@ -3017,12 +3029,34 @@ impl IssueNftFixture {
+            nft_metadata: self.params.nft_metadata.clone(),
+            non_channel_owner: self.params.non_channel_owner.clone(),
+            royalty: Some(royalty_pct),
+        };
```
```suggestion
        let new_params = NftIssuanceParameters::<Test> {
            royalty: Some(royalty_pct),
            ...self.params.clone()
        };
```

Lezek123 commented on runtime-modules/membership/Cargo.toml:
```diff
@@ -6,24 +6,24 @@ edition = '2018'
 
 [dependencies]
```
Version of this package (`runtime-modules/membership`) was not updated

mnaamani replied:
will update, I mistakenly assumed v5.0.0 was new version on audit3 turns out its same as on master.

Lezek123 commented on runtime-modules/project-token/Cargo.toml:
```diff
@@ -1,33 +1,35 @@
 [package]
 name = "pallet-project-token"
-version = '0.1.0'
+version = '5.0.0'
```
Incorrect update from `0.1.0` to `5.0.0`.
```suggestion
version = '1.0.0'
```
_I suggest sticking to `1.0.0`, since we never actually used this module on any testnet yet_

mnaamani replied:
That makes sense of course, but since we never publish anything to a remote crates registry or use our crates in other projects, it doesn't really make a difference :) But I will commit suggestion

mnaamani replied:
updated in 27f693078d3a90bd676a3f774f4b5ca837382b99

Lezek123 commented on runtime-modules/project-token/src/types.rs:
```diff
@@ -1158,6 +1200,7 @@ where
             sale: None,
             transfer_policy: params.transfer_policy.into(),
             patronage_info,
+            // in order to avoid underflow when Token is started idle and `update_upcoming_sale` is called`
```
Leftover comment, can now be removed

mnaamani replied:
removed in 27f693078d3a90bd676a3f774f4b5ca837382b99

Lezek123 commented on runtime-modules/proposals/codex/src/tests/mock.rs:
```diff
@@ -1,79 +1,350 @@
+        Timestamp: pallet_timestamp::{Pallet, Call, Storage, Inherent},
+        Balances: balances::{Pallet, Call, Storage, Config<T>, Event<T>},
+        Staking: staking::{Pallet, Call, Config<T>, Storage, Event<T>},
+        Session: pallet_session::{Pallet, Call, Storage, Event, Config<T>},
```
Why do we actually need to mock `Session`, `Authorship` and `Historical` here?
It adds a lot of complexity and `Session` is only really used by `is_disabled` function, which itself is never used.
If I replace:
```
    type SessionInterface = Self;
    type NextNewSession = Session;
```
with
```
    type SessionInterface = ();
    type NextNewSession = ();
```
in `impl staking::Config for Test` all tests seem to still work as expected and I can comment out the implementation of the pallets mentioned above.

mnaamani replied:
They are needed if you run the benchmark tests. Can you confirm.

mnaamani replied:
`cargo test --release -p pallet-proposals-codex --features runtime-benchmarks`

Lezek123 replied:
I can successfully execute `cargo test --release -p pallet-proposals-codex --features runtime-benchmarks` without them

mnaamani replied:
Great I also confirmed running the weight generation works, because of-course the runtime used when generating weights is the "real" runtime, not the mocked one in the tests. :+1:
Will drop those pallets from the test mocks.

Lezek123 commented on runtime-modules/proposals/engine/src/tests/mock/mod.rs:
```diff
@@ -436,46 +409,55 @@ impl crate::VotersParameters for () {
-    type MaximumBlockLength = MaximumBlockLength;
-    type AvailableBlockRatio = AvailableBlockRatio;
+    type Event = Event;
+    type BlockHashCount = ConstU64<250>;
```
It's redundant to have `pub const BlockHashCount: u64 = 250;` parameter defined if we're using `ConstU64<250>` here anyway

mnaamani replied:
Yes this is probably showing up as a clippy warning which I haven't addressed yet. I'm sure there are more of these.

Lezek123 commented on runtime-modules/staking-handler/Cargo.toml:
```diff
@@ -1,23 +1,25 @@
 [package]
 name = 'pallet-staking-handler'
-version = '2.0.0'
+version = '5.0.0'
```
Version incorrectly updated
```suggestion
version = '3.0.0'
```

Lezek123 commented on runtime-modules/storage/src/tests/fixtures.rs:
```diff
@@ -1042,7 +1047,8 @@ pub struct SetStorageBucketVoucherLimitsFixture {
 impl SetStorageBucketVoucherLimitsFixture {
     pub fn call_and_assert(&self, expected_result: DispatchResult) {
-        let old_voucher = Storage::storage_bucket_by_id(self.storage_bucket_id).voucher;
+        let old_voucher = Storage::storage_bucket_by_id(self.storage_bucket_id);
```
Variable name is incorrect - this is `Option<StorageBucket>`, not a voucher

Lezek123 commented on runtime-modules/storage/src/tests/fixtures.rs:
```diff
@@ -1051,14 +1057,14 @@ impl SetStorageBucketVoucherLimitsFixture {
 
         assert_eq!(actual_result, expected_result);
-        let new_voucher = Storage::storage_bucket_by_id(self.storage_bucket_id).voucher;
+        let new_voucher = Storage::storage_bucket_by_id(self.storage_bucket_id);
```
Variable name is incorrect - this is `Option<StorageBucket>`, not a voucher

Lezek123 commented on runtime-modules/utility/Cargo.toml:
```diff
@@ -1,32 +1,33 @@
 [package]
 name = 'pallet-utility'
-version = '1.1.0'
+version = '5.0.0'
```
Incorrect version update
```suggestion
version = '2.0.0'
```

Lezek123 commented on runtime-modules/utility/src/tests/mocks.rs:
```diff
@@ -166,36 +117,92 @@ pub type OperationsWorkingGroupInstanceGamma = working_group::Instance8;
-    type MaximumBlockLength = MaximumBlockLength;
-    type AvailableBlockRatio = AvailableBlockRatio;
+    type Event = Event;
+    type BlockHashCount = ConstU64<250>;
```
As pointed out in one of the previous comments: It doesn't make much sense to have both `pub const BlockHashCount: u64 = 250;` defined and use `ConstU64<250>` here. Only one of the approaches should be chosen.

Lezek123 commented on runtime-modules/common/Cargo.toml:
```diff
@@ -8,14 +8,15 @@ edition = '2018'
 serde = { version = "1.0.101", optional = true, features = ["derive"] }
 strum = {version = "0.19", optional = true}
 strum_macros = {version = "0.19", optional = true}
-codec = { package = 'parity-scale-codec', version = '1.3.4', default-features = false, features = ['derive'] }
```
Version of this package (`runtime-modules/common`) was not updated

Lezek123 commented on runtime-modules/constitution/Cargo.toml:
```diff
@@ -6,18 +6,18 @@ edition = '2018'
 
 [dependencies]
 serde = { version = "1.0.101", optional = true, features = ["derive"] }
-codec = { package = 'parity-scale-codec', version = '1.3.1', default-features = false, features = ['derive'] }
```
Version of this package (`runtime-modules/constitution`) was not updated

Lezek123 commented on runtime-modules/council/Cargo.toml:
```diff
@@ -5,27 +5,28 @@ authors = ['Joystream contributors']
 edition = '2018'
 
 [dependencies]
-codec = { package = 'parity-scale-codec', version = '1.0.0', default-features = false, features = ['derive'] }
```
Version of this package (`runtime-modules/council`) was not updated

Lezek123 commented on runtime-modules/forum/Cargo.toml:
```diff
@@ -6,26 +6,27 @@ edition = '2018'
 
 [dependencies]
 serde = { version = "1.0.101", optional = true, features = ["derive"] }
-codec = { package = 'parity-scale-codec', version = '1.3.4', default-features = false, features = ['derive'] }
```
Version of this package (`runtime-modules/forum`) was not updated

Lezek123 commented on runtime-modules/blog/Cargo.toml:
```diff
@@ -5,36 +5,41 @@ authors = ['Joystream contributors']
 edition = '2018'
 
 [dependencies]
-sp-std = { package = 'sp-std', default-features = false, git = 'https://github.com/paritytech/substrate.git', rev = '2cd20966cc09b059817c3ebe12fc130cdd850d62'}
```
Version of this package (`runtime-modules/blog`) was not updated

Lezek123 commented on runtime-modules/proposals/codex/Cargo.toml:
```diff
@@ -6,15 +6,16 @@ edition = '2018'
 
 [dependencies]
 serde = { version = "1.0.101", optional = true, features = ["derive"] }
-codec = { package = 'parity-scale-codec', version = '1.3.4', default-features = false, features = ['derive'] }
```
Version of this package (`runtime-modules/proposals/codex`) was not updated

Lezek123 commented on runtime-modules/proposals/discussion/Cargo.toml:
```diff
@@ -6,24 +6,25 @@ edition = '2018'
 
 [dependencies]
 serde = { version = "1.0.101", optional = true, features = ["derive"] }
-codec = { package = 'parity-scale-codec', version = '1.3.4', default-features = false, features = ['derive'] }
```
Version of this package (`runtime-modules/proposals/discussion`) was not updated

Lezek123 commented on runtime-modules/proposals/engine/Cargo.toml:
```diff
@@ -6,26 +6,27 @@ edition = '2018'
 
 [dependencies]
 serde = { version = "1.0.101", optional = true, features = ["derive"] }
-codec = { package = 'parity-scale-codec', version = '1.3.4', default-features = false, features = ['derive'] }
```
Version of this package (`runtime-modules/proposals/engine`) was not updated

Lezek123 commented on runtime-modules/referendum/Cargo.toml:
```diff
@@ -5,26 +5,27 @@ authors = ['Joystream contributors']
 edition = '2018'
 
 [dependencies]
-codec = { package = 'parity-scale-codec', version = '1.0.0', default-features = false, features = ['derive'] }
```
Version of this package (`runtime-modules/referendum`) was not updated

mnaamani committed:
github workflow - run exit status seems to only come from last command

Lezek123 reviewed (changes requested):
1. `cargo build --release --features runtime-benchmarks` still fails for me with:
```
  error: duplicate lang item in crate `std` (which `pallet_election_provider_support_benchmarking` depends on): `panic_impl`.
    |
    = note: the lang item is first defined in crate `sp_io` (which `joystream_node_runtime` depends on)
    = note: first definition in `sp_io` loaded from /home/leszek/projects/joystream/joystream-ws-4/target/release/wbuild/joystream-node-runtime/target/wasm32-unknown-unknown/release/deps/libsp_io-36acee912839bf14.rmeta
    = note: second definition in `std` loaded from /home/leszek/.rustup/toolchains/nightly-2022-05-11-x86_64-unknown-linux-gnu/lib/rustlib/wasm32-unknown-unknown/lib/libstd-5a0663610dd354de.rlib

  error: duplicate lang item in crate `std` (which `pallet_election_provider_support_benchmarking` depends on): `oom`.
    |
    = note: the lang item is first defined in crate `sp_io` (which `joystream_node_runtime` depends on)
    = note: first definition in `sp_io` loaded from /home/leszek/projects/joystream/joystream-ws-4/target/release/wbuild/joystream-node-runtime/target/wasm32-unknown-unknown/release/deps/libsp_io-36acee912839bf14.rmeta
    = note: second definition in `std` loaded from /home/leszek/.rustup/toolchains/nightly-2022-05-11-x86_64-unknown-linux-gnu/lib/rustlib/wasm32-unknown-unknown/lib/libstd-5a0663610dd354de.rlib

```
2. `export RUNTIME_PROFILE="TESTING" && ./scripts/cargo-build.sh` also fails:
```
  error[E0425]: cannot find function `update_global_nft_limit_proposal` in this scope
     --> /home/leszek/projects/joystream/joystream-ws-4/runtime/src/proposals_configuration/mod.rs:104:9
      |
  104 |         update_global_nft_limit_proposal();
      |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in this scope


  error: duplicate lang item in crate `std`: `panic_impl`.
    |
    = note: the lang item is first defined in crate `sp_io` (which `joystream_node_runtime` depends on)
    = note: first definition in `sp_io` loaded from /home/leszek/projects/joystream/joystream-ws-4/target/release/wbuild/joystream-node-runtime/target/wasm32-unknown-unknown/release/deps/libsp_io-36acee912839bf14.rmeta
    = note: second definition in `std` loaded from /home/leszek/.rustup/toolchains/nightly-2022-05-11-x86_64-unknown-linux-gnu/lib/rustlib/wasm32-unknown-unknown/lib/libstd-5a0663610dd354de.rlib

  error: duplicate lang item in crate `std`: `oom`.
    |
    = note: the lang item is first defined in crate `sp_io` (which `joystream_node_runtime` depends on)
    = note: first definition in `sp_io` loaded from /home/leszek/projects/joystream/joystream-ws-4/target/release/wbuild/joystream-node-runtime/target/wasm32-unknown-unknown/release/deps/libsp_io-36acee912839bf14.rmeta
    = note: second definition in `std` loaded from /home/leszek/.rustup/toolchains/nightly-2022-05-11-x86_64-unknown-linux-gnu/lib/rustlib/wasm32-unknown-unknown/lib/libstd-5a0663610dd354de.rlib
```

mnaamani committed:
add missing proposal config for ntf limits for staging and testing profiles

mnaamani commented:
1. Yes I haven't tried to fix the runtime benchmarks build yet

2. Thanks for the re-review. I did in fact notice why the build was failing for testing and staging profile, I was trying to find out why it didn't properly fail in CI.. seems the job doesn't behave as expected. I was expecting behavior like `set -e` so if you have:
```
run: |
  yarn cargo-checks && yarn cargo-build
  ./target/release/joystream-node --version
```

I was expecting job to fail if checks or build exits with non zero.
But it continued. So I just made the last line be the `yarn cargo-checks && yarn cargo-build`

mnaamani committed:
Update runtime-modules/content/src/tests/fixtures.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

mnaamani committed:
Update runtime-modules/content/src/tests/fixtures.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

mnaamani committed:
Update runtime-modules/content/src/tests/fixtures.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

mnaamani committed:
update project-token version to 1.0.0

mnaamani committed:
correct pallet versions

mnaamani committed:
fix var name in fixture

mnaamani committed:
typo in content module

mnaamani committed:
Cargo.lock

mnaamani committed:
fix benchmark build - still need to drop blog

mnaamani committed:
drop blog pallet from benchmarks dependencies

mnaamani committed:
make benchmark command work without benchmark metadata

mnaamani committed:
make note of importance of order of Authorship module

mnaamani committed:
patch crates.io to use joystream ss58-registry crate

mnaamani committed:
cargo-fmt

mnaamani committed:
consistent working group instances across mock and runtime

mnaamani committed:
note about benchmark metadata

mnaamani committed:
correct comment

mnaamani committed:
fix comment

mnaamani committed:
fix generate weights script

dobertRowneySr reviewed (approved):
LGTM, I have just one question:
why don't we update content pallet version?

mnaamani commented:
> LGTM, I have just one question: why don't we update content pallet version?

Good Idea, maybe I missed it :)

mnaamani committed:
update content pallet version

mnaamani commented:
I'm pretty much done with this PR. I've made more progress on benchmarks build and weigh generation scripts.. but I will open a separate PR to make it clearer what I've changed.

Lezek123 commented on runtime-modules/storage/src/tests/mocks.rs:
```diff
@@ -53,7 +53,7 @@ frame_support::construct_runtime!(
         // Need to be added for benchmarks to work
         Wg2: working_group::<Instance2>::{Pallet, Call, Storage, Event<T, I>},
-        Wg3: working_group::<Instance3>::{Pallet, Call, Storage, Event<T, I>},
+        Wg3: working_group::<Instance9>::{Pallet, Call, Storage, Event<T, I>},
```
```suggestion
        Wg9: working_group::<Instance9>::{Pallet, Call, Storage, Event<T, I>},
```

mnaamani replied:
will fix

Lezek123 commented on Cargo.toml:
```diff
@@ -29,3 +29,6 @@ exclude = [
 panic = "unwind"
+
+[patch.crates-io]
+ss58-registry = { pacakge = 'ss58-registry', git = 'https://github.com/bwhm/ss58-registry', rev = 'a70121e7bbb56f983f6f17771cc79d8f19b481a0'}
```
When building the runtime I'm getting:
```
warning: /home/leszek/projects/joystream/joystream-ws-4/Cargo.toml: unused manifest key: patch.crates-io.ss58-registry.pacakge
```

mnaamani replied:
oops typo, `pacakge` is misspelled, should be `package`

Lezek123 reviewed (commented):
1. There are some remaining comments not addressed from my first review (https://github.com/Joystream/joystream/pull/3928#pullrequestreview-1013292575), but it's nothing major, namely:
1.1. missing `CHANGELOG` update
1.2. duplicate dependency in [runtime-modules/storage/Cargo.toml](https://github.com/Joystream/joystream/pull/3928/files/0a4e8f411e5a94dc4c2f926639a66e5686ac1764#diff-b0d88d068428418a5d044fb89b27548c36f0c65abe6df8d076fdbe9e917b8aca)
2. I encountered some strange errors in benchmarking tests:
```
failures:

---- benchmarking::tests::test_reveal_vote_space_replace_last_winner stdout ----
thread 'benchmarking::tests::test_reveal_vote_space_replace_last_winner' panicked at 'assertion failed: `(left == right)`
  left: `Revealing(ReferendumStageRevealing { started: 5, winning_target_count: 10, intermediate_winners: [OptionResult { option_id: 1, vote_power: 20003 }, OptionResult { option_id: 0, vote_power: 20002 }, OptionResult { option_id: 2, vote_power: 10001 }, OptionResult { option_id: 3, vote_power: 10001 }, OptionResult { option_id: 4, vote_power: 10001 }, OptionResult { option_id: 5, vote_power: 10001 }, OptionResult { option_id: 6, vote_power: 10001 }, OptionResult { option_id: 7, vote_power: 10001 }, OptionResult { option_id: 8, vote_power: 10001 }, OptionResult { option_id: 9, vote_power: 10001 }], current_cycle_id: 0 })`,
 right: `Revealing(ReferendumStageRevealing { started: 5, winning_target_count: 10, intermediate_winners: [OptionResult { option_id: 0, vote_power: 10001 }, OptionResult { option_id: 1, vote_power: 10001 }, OptionResult { option_id: 2, vote_power: 10001 }, OptionResult { option_id: 3, vote_power: 10001 }, OptionResult { option_id: 4, vote_power: 10001 }, OptionResult { option_id: 5, vote_power: 10001 }, OptionResult { option_id: 6, vote_power: 10001 }, OptionResult { option_id: 7, vote_power: 10001 }, OptionResult { option_id: 8, vote_power: 10001 }, OptionResult { option_id: 9, vote_power: 10001 }], current_cycle_id: 0 })`: Votes not revealed', runtime-modules/referendum/src/benchmarking.rs:356:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- benchmarking::tests::test_reveal_vote_space_not_in_winners stdout ----
thread 'benchmarking::tests::test_reveal_vote_space_not_in_winners' panicked at 'assertion failed: `(left == right)`
  left: `Revealing(ReferendumStageRevealing { started: 5, winning_target_count: 10, intermediate_winners: [OptionResult { option_id: 0, vote_power: 20002 }, OptionResult { option_id: 1, vote_power: 20002 }, OptionResult { option_id: 2, vote_power: 10001 }, OptionResult { option_id: 3, vote_power: 10001 }, OptionResult { option_id: 4, vote_power: 10001 }, OptionResult { option_id: 5, vote_power: 10001 }, OptionResult { option_id: 6, vote_power: 10001 }, OptionResult { option_id: 7, vote_power: 10001 }, OptionResult { option_id: 8, vote_power: 10001 }, OptionResult { option_id: 9, vote_power: 10001 }], current_cycle_id: 0 })`,
 right: `Revealing(ReferendumStageRevealing { started: 5, winning_target_count: 10, intermediate_winners: [OptionResult { option_id: 0, vote_power: 10001 }, OptionResult { option_id: 1, vote_power: 10001 }, OptionResult { option_id: 2, vote_power: 10001 }, OptionResult { option_id: 3, vote_power: 10001 }, OptionResult { option_id: 4, vote_power: 10001 }, OptionResult { option_id: 5, vote_power: 10001 }, OptionResult { option_id: 6, vote_power: 10001 }, OptionResult { option_id: 7, vote_power: 10001 }, OptionResult { option_id: 8, vote_power: 10001 }, OptionResult { option_id: 9, vote_power: 10001 }], current_cycle_id: 0 })`: Votes not revealed', runtime-modules/referendum/src/benchmarking.rs:356:5

---- benchmarking::tests::test_reveal_vote_space_for_new_winner stdout ----
thread 'benchmarking::tests::test_reveal_vote_space_for_new_winner' panicked at 'assertion failed: `(left == right)`
  left: `Revealing(ReferendumStageRevealing { started: 5, winning_target_count: 10, intermediate_winners: [OptionResult { option_id: 0, vote_power: 20003 }, OptionResult { option_id: 1, vote_power: 10001 }, OptionResult { option_id: 2, vote_power: 10001 }, OptionResult { option_id: 3, vote_power: 10001 }, OptionResult { option_id: 4, vote_power: 10001 }, OptionResult { option_id: 5, vote_power: 10001 }, OptionResult { option_id: 6, vote_power: 10001 }, OptionResult { option_id: 7, vote_power: 10001 }, OptionResult { option_id: 8, vote_power: 10001 }], current_cycle_id: 0 })`,
 right: `Revealing(ReferendumStageRevealing { started: 5, winning_target_count: 10, intermediate_winners: [OptionResult { option_id: 0, vote_power: 10001 }, OptionResult { option_id: 1, vote_power: 10001 }, OptionResult { option_id: 2, vote_power: 10001 }, OptionResult { option_id: 3, vote_power: 10001 }, OptionResult { option_id: 4, vote_power: 10001 }, OptionResult { option_id: 5, vote_power: 10001 }, OptionResult { option_id: 6, vote_power: 10001 }, OptionResult { option_id: 7, vote_power: 10001 }, OptionResult { option_id: 8, vote_power: 10001 }], current_cycle_id: 0 })`: Votes not revealed', runtime-modules/referendum/src/benchmarking.rs:356:5

---- benchmarking::tests::test_reveal_vote_already_existing stdout ----
thread 'benchmarking::tests::test_reveal_vote_already_existing' panicked at 'assertion failed: `(left == right)`
  left: `Revealing(ReferendumStageRevealing { started: 5, winning_target_count: 10, intermediate_winners: [OptionResult { option_id: 0, vote_power: 30003 }, OptionResult { option_id: 1, vote_power: 10001 }, OptionResult { option_id: 2, vote_power: 10001 }, OptionResult { option_id: 3, vote_power: 10001 }, OptionResult { option_id: 4, vote_power: 10001 }, OptionResult { option_id: 5, vote_power: 10001 }, OptionResult { option_id: 6, vote_power: 10001 }, OptionResult { option_id: 7, vote_power: 10001 }, OptionResult { option_id: 8, vote_power: 10001 }, OptionResult { option_id: 9, vote_power: 10001 }], current_cycle_id: 0 })`,
 right: `Revealing(ReferendumStageRevealing { started: 5, winning_target_count: 10, intermediate_winners: [OptionResult { option_id: 0, vote_power: 10001 }, OptionResult { option_id: 1, vote_power: 10001 }, OptionResult { option_id: 2, vote_power: 10001 }, OptionResult { option_id: 3, vote_power: 10001 }, OptionResult { option_id: 4, vote_power: 10001 }, OptionResult { option_id: 5, vote_power: 10001 }, OptionResult { option_id: 6, vote_power: 10001 }, OptionResult { option_id: 7, vote_power: 10001 }, OptionResult { option_id: 8, vote_power: 10001 }, OptionResult { option_id: 9, vote_power: 10001 }], current_cycle_id: 0 })`: Votes not revealed', runtime-modules/referendum/src/benchmarking.rs:356:5


failures:
    benchmarking::tests::test_reveal_vote_already_existing
    benchmarking::tests::test_reveal_vote_space_for_new_winner
    benchmarking::tests::test_reveal_vote_space_not_in_winners
    benchmarking::tests::test_reveal_vote_space_replace_last_winner
```
But I suppose it's ok if those are to be addressed in the next PR

mnaamani committed:
update runtime CHANGELOG

mnaamani committed:
storage mocks fix working group instance name

mnaamani committed:
typo in Cargo.toml

mnaamani committed:
simplify proposals codex mocks - drop session, historical and authorship modules

mnaamani commented:
@Lezek123 I pushed final fixes. Kept the changelog very brief, I'm not clear on the deeper changes made so perhaps we can update it before release. The remaining points (broken benchmark test for referendum and duplicate dependency will be addressed in another PR)

Lezek123 reviewed (approved):
Great stuff, I think we're ready for Substrate v3!
